### PR TITLE
Sharktank LLM split argmax op

### PR DESCRIPTION
**Pull request by stbaione**
Putting up a draft PR to get feedback on my attempt at manual `split_argmax`.

I was seeing worse performance metrics when benchmarking it:

- Argmax: `~4ms`
- Split Argmax: `~4.5ms`

So, I'm assuming that the parallelization isn't happening the way that we want it to. Wanted to get some feedback on the implementation, there's a pretty high likelihood that I missed a concept that would allow us to actually achieve the parallelization.

### CODE REVIEW

**Review by vinayakdsci — COMMENTED** [review]
cc @rsuderman 

From the conversation I had with @MaheshRavishankar when the kernel performance metric initially came up, this is equivalent to how the compiler would handle it eventually by split reducing the kernel.

```python
def argmax(result_logits: sfnp.device_array):
    # Reshape the result logits first into (current_ndim + 1) dimensions.
    # The reshaped innermost dimension will need choosing based on the vocab size
    # and the way you would want to distribute.
    result_logits_reshaped = result_logits.reshape(*result_logits.shape[:-1], result_logits.shape[-1] // 1024, 1024) 
    argmax_1 = torch.argmax(result_logits_reshaped, dim=-1)
    # The innermost dim should now be reduced. Do another reshape (this can possibly be removed if that dimension does not show up).
    reshape_2 = argmax_1.reshape(*result_logits.shape[:-1], result_logits.shape[-1] // 1024)
    # This reduces the innermost dimension again.
    final_argmax = torch.argmax(reshape_2, dim=-1)
    return final_argmax
```
This implementation ran, but had numerics issues. The implementation does follow the idea of split reduction correctly, and this would be the correct way if the results are accurate.

File: sharktank/sharktank/ops/signatures.py [hunk 1]
```diff
@@ -749,6 +758,32 @@ def _matmul_trampoline(
         d.fail(tensors)
 
 
+@overridable
+def max(
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/signatures.py]: Removed those ops

**Comment by stbaione** [comment]
> cc @rsuderman
> 
> From the conversation I had with @MaheshRavishankar when the kernel performance metric initially came up, this is equivalent to how the compiler would handle it eventually by split reducing the kernel.
> 
> ```python
> def argmax(result_logits: sfnp.device_array):
>     # Reshape the result logits first into (current_ndim + 1) dimensions.
>     # The reshaped innermost dimension will need choosing based on the vocab size
>     # and the way you would want to distribute.
>     result_logits_reshaped = result_logits.reshape(*result_logits.shape[:-1], result_logits.shape[-1] // 1024, 1024) 
>     argmax_1 = torch.argmax(result_logits_reshaped, dim=-1)
>     # The innermost dim should now be reduced. Do another reshape (this can possibly be removed if that dimension does not show up).
>     reshape_2 = argmax_1.reshape(*result_logits.shape[:-1], result_logits.shape[-1] // 1024)
>     # This reduces the innermost dimension again.
>     final_argmax = torch.argmax(reshape_2, dim=-1)
>     return final_argmax
> ```
> 
> This implementation ran, but had numerics issues. The implementation does follow the idea of split reduction correctly, and this would be the correct way if the results are accurate.

Pushed a new implementation that follows this idea and passes the tests numerically

File: sharktank/tests/ops/ops_test.py [hunk 2]
```diff
@@ -42,6 +42,39 @@ def testArgmaxDim0(self):
...
+            a[1][2] = 99
+            a[0][1] = 1
+            a[0][3] = 1
+            result = ops.argmax(a, dim=0, keepdim=True)
+            expected = torch.tensor([[1, 0, 1, 0]], dtype=torch.int64)
+            assert result.shape == (1, 4)
+            assert torch.equal(result, expected)
+
+
+class SplitArgmaxTest(unittest.TestCase):
```
Comment (Alex-Vasile) [inline sharktank/tests/ops/ops_test.py:58]: I think these tests should go under `ArgmaxTest` since this is a still argmax. See for example `MatmulTest` which has the custom and default versions in the same class.

File: sharktank/tests/ops/ops_test.py [hunk 3]
```diff
@@ -42,6 +42,39 @@ def testArgmaxDim0(self):
...
+    def testSplitArgmax(self):
+        for dtype in [torch.float16, torch.float32]:
+            a = torch.zeros(1, 1, 256, dtype=dtype)
+            a[0][0][42] = 42
+            assert ops.custom_impls.split_argmax(a, -1, 16) == 42
```
Comment (Alex-Vasile) [inline sharktank/tests/ops/ops_test.py:63]: Use a @parameterized instead of a for-loop inside the test. Makes it easier to debug, and lets them run in parallel.

Comment (Alex-Vasile) [reply sharktank/tests/ops/ops_test.py:63]: Same comment for all of the other tests with the for loop.

**Review by Alex-Vasile — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/ops/custom_impls.py [hunk 4]
```diff
@@ -48,6 +48,33 @@
 #     return mmtfp(lhs, rhs)
 
 
+# Argmax
+
+
+@argmax.override(Tensor, QuantizedTensor)
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/custom_impls.py]: Fix override to just say Tensor, was under the impression that this just denoted the types of Tensors supported for the op, as opposed to a tuple of expected arg types

File: sharktank/sharktank/ops/signatures.py [hunk 5]
```diff
@@ -125,16 +125,23 @@ def _all_reduce_trampoline(d: SignatureDispatcher, tensor: AnyTensor):
 
 
 @overridable
-def argmax(tensor: AnyTensor, axis: int) -> AnyTensor:
+def argmax(
+    tensor: AnyTensor, dim: Optional[int] = None, keepdim: bool = False
+) -> AnyTensor:
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/signatures.py]: Fixed

File: sharktank/tests/ops/ops_test.py (hunk 3 above)
Comment (stbaione) [reply sharktank/tests/ops/ops_test.py:61]: Combined tests into one class, but `pytest.mark.parametrize` decorator is incompatible with `unittest.TestCase`, which is what the rest of the tests are written with:

[source](https://stackoverflow.com/questions/18182251/does-pytest-parametrized-test-work-with-unittest-class-based-tests)

File: sharktank/sharktank/ops/signatures.py [hunk 6]
```diff
@@ -125,16 +125,23 @@ def _all_reduce_trampoline(d: SignatureDispatcher, tensor: AnyTensor):
...
     ...
 
 
 @argmax.trampoline
-def _argmax_trampoline(d: SignatureDispatcher, tensor: AnyTensor, axis: int):
+def _argmax_trampoline(
+    d: SignatureDispatcher,
+    tensor: AnyTensor,
+    dim: Optional[int] = None,
+    keepdim: bool = False,
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/signatures.py:144]: Fixed

File: sharktank/tests/ops/ops_test.py (hunk 2 above)
Comment (stbaione) [reply sharktank/tests/ops/ops_test.py]: Combined into a single test class

File: sharktank/sharktank/ops/default_impls.py [hunk 7]
```diff
@@ -27,6 +27,15 @@
 import iree.turbine.ops.iree
 
 
+@argmax.override(Tensor)
+def argmax_default(
+    x: Tensor,
+    dim: Optional[int] = None,
+    keepdim: bool = False,
+) -> None:
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/default_impls.py:35]: Moving switch from `signatures` to `default_impl`

[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/default_impls.py:36]: Moved over to `default_impl`

File: sharktank/sharktank/ops/custom_impls.py [hunk 8]
```diff
@@ -48,6 +48,33 @@
...
+def split_argmax(input_tensor, dim, chunk_size: int = 128):
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/ops/custom_impls.py]: Added `keepdim`

File: sharktank/sharktank/examples/export_paged_llm_v1.py [hunk 9]
```diff
@@ -435,11 +435,17 @@ def _(
             return logits
 
     def generate_argmax():
+        # TODO: Remove this when the corresponding `dtype` conversion is
+        # removed in `PagedLlmModelV1.prefill/decode`
```
Comment (Alex-Vasile) [inline sharktank/sharktank/examples/export_paged_llm_v1.py:439]: Is someone actively working on that? If not, it may be worth creating an issue to track this.

File: sharktank/sharktank/ops/default_impls.py [hunk 10]
```diff
@@ -27,6 +27,47 @@
 import iree.turbine.ops.iree
 
 
+@argmax.override(Tensor)
+def argmax_default(
+    x: Tensor,
+    dim: Optional[int] = None,
+    chunk_size: Optional[int] = None,
+    keepdim: bool = False,
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/default_impls.py:35]: These two should be flipped. `chunk_size` is not an argument for `torch.argmax`, and may lead to silent bugs if people call `ops.argmax` expecting positional arguments to be in the same order as `torch.argmax` if torch will accept a bool instead of an int. Same idea for other signatures for this function.

File: sharktank/sharktank/ops/default_impls.py [hunk 11]
```diff
@@ -27,6 +27,47 @@
...
+    argmax_1 = argmax(tensor_split, dim + 1)
+    argmax_expanded = unsqueeze(argmax_1, dim + 1)
+
+    max_vals = gather(tensor_split, dim + 1, argmax_expanded)
+    max_vals = squeeze(max_vals, dim + 1)
+
+    argmax_2 = argmax(max_vals, dim)
+    argmax_2_expanded = unsqueeze(argmax_2, 0)
+
+    final_index_in_chunk = gather(argmax_1, dim, argmax_2_expanded)
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/default_impls.py:60]: Shouldn't this use argmax_1_expanded?

**Review by Alex-Vasile — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/examples/export_paged_llm_v1.py (hunk 9 above)
Comment (stbaione) [reply sharktank/sharktank/examples/export_paged_llm_v1.py:439]: It's a potential result of us moving the Shortfin server over to GPU kernels. Right now, we cast the logits in the model to fp16 for those formats, so that we have better support and performance in things like `xtensor, numpy, etc.`.

If we move the LLM server over the GPU kernels completely, then we may be able to remove that, should we see a potential perf benefit from it.

It may remain though, because we might want to be able to swap between CPU and GPU invocations (i.e. if a user doesn't have a GPU).

So, I'm not sure if it's an issue per-say, but something I wanted to make sure was documented in case we remove that in the future

File: sharktank/sharktank/ops/default_impls.py (hunk 11 above)
Comment (rsuderman) [reply sharktank/sharktank/ops/default_impls.py:60]: Nope. We are just gathering the values, having the additional dimension doesn't matter.

File: sharktank/sharktank/ops/signatures.py [hunk 12]
```diff
@@ -125,16 +125,27 @@ def _all_reduce_trampoline(d: SignatureDispatcher, tensor: AnyTensor):
 
 
 @overridable
-def argmax(tensor: AnyTensor, axis: int) -> AnyTensor:
+def argmax(
+    tensor: AnyTensor,
+    dim: Optional[int] = None,
+    chunk_size: Optional[int] = None,
```
Comment (rsuderman) [inline sharktank/sharktank/ops/signatures.py:131]: You will need to update the `sharded_impls.py` version as well so that it has `chunk_size`. It should just involve proxy-ing the value to the sub calls.

**Review by rsuderman — CHANGES_REQUESTED** [review]

**Review by rsuderman — APPROVED** [review]
Wait for presubmits but this should be good to land.

**Review by Alex-Vasile — APPROVED** [review]
